Add album_level_tags plugin

Harakku opened the pull request:
New plugin to transform multiple file specific tags into single
strings which are usable in the file naming script via script
functions.

zas commented on plugins/album_level_tags/album_level_tags.py:
```diff
@@ -0,0 +1,360 @@
+# START OF MONKEYPATCHED HOOKS
+#
+#
+
```
In fact, it would be great to submit PRs to add those hooks to the core Picard code, so they'll be in next stable version. There's already a PR for a post file save hook (https://github.com/metabrainz/picard/pull/1186).

Harakku replied:
I did that, removed the now unneeded code from the plugin and tested that it still works the same as before. Links to PRs: https://github.com/metabrainz/picard/pull/1190, https://github.com/metabrainz/picard/pull/1191

I bumped the needed plugin API version to "2.3" since the plugin doesn't work with the current builds.

I also read more about licensing and concluded that since the plugin imports functions from Picard, the plugin needs to be GPL2 too. No big deal though.

Harakku committed:
Add album_level_tags plugin

New plugin to transform multiple file specific tags into single
strings which are usable in the file naming script via script
functions.

sambhav commented on plugins/album_level_tags/album_level_tags.py:
```diff
@@ -0,0 +1,290 @@
+        return
+
+    try:
+        tag_variable[album_mbid][metadata_value] -= 1
```
This logic and the one above might be simplified using either collections.Counter or collections.defaultdict.

Harakku replied:
That made a big difference and those changes kinda snowballed into making everything else cleaner too. :-)

There are now some formatting options in the script functions. Based on my manual testing, everything seems to work.

zas reviewed (approved):
LGTM

@phw: ok for you ?

Harakku commented:
I just realized how this plugin could be made much cleaner. Instead of scribing down tags as they come by, an album object can be saved as a global variable. Then the album object can be accessed when the script functions are being called. Duhhh.

Gonna make those changes soonish.

zas commented:
> Gonna make those changes soonish.

Should we merge as is or do you have changes to make before?

Harakku commented:
Almost forgot to update this. The plugin has some option pages now and quite big changes in the way it works. Didn't end up needing the new hooks after all. (Awkward.) The option editing is a bit rough but I hope it's somewhat easy to understand.

All in all, I'm happy with the plugin now.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+
+        # Check if cached data is recent enough to be used (i.e. we can assume that the user hasn't modified the tags
+        # in the meantime) or remove other releases from the cache to keep it small
+        global tag_cache
```
`global` keyword isn't strictly needed here I think.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+            t = time.time()
+
+            cached_results = album_cache.get((metadata_key, option_profile))
+            if cached_results and t - last_run_time < 4:  # seconds
```
I'd replace `4` with a explicit variable/constant (something like `DELAY_IN_SECONDS`) and remove the comment.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+        global last_run_time
+        album_cache = tag_cache.get(album.id, {})
+        if album_cache:
+            t = time.time()
```
Instead of `t`, I'd use `now` or `current_time` since it stores current time.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+
+        if option_profile is not None and option_profile not in all_user_options:
+            raise KeyError("{0}: Unknown option profile \"{1}\" used in a script.".format(PLUGIN_NAME, option_profile))
+        if profile_options.get(FORMATTER_WHEN_ONE_OPTION):
```
Perhaps:
```python
formatter2_is_default = False if profile_options.get(FORMATTER_WHEN_ONE_OPTION) else True
```
Since you're using few of those below.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+        group_by_medium = True if metadata_key == "media" else False
+        discnumber = 0
+        temp = defaultdict(list)
+        for t in album.tracks:
```
Use `track` instead of `t`.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+
+        if average_mode:
+            tag_counter = Counter({ sum(map(float, tag_list)) / len(tag_list): 1
+                                    for tag_list in temp.values() if len(tag_list) > 0 })
```
This expression deserves at least a comment, you may also cache `len(tag_list)`.

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+
+        # Effectively removes the None key from the counter
+        tag_counter[None] = 0
+        tag_counter = +tag_counter
```
Typo? Or?

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+        text_area = self.ui.plainTextEdit_raw_2
+        label = self.ui.label_valid_raw_2
+        try:
+            _ = json.loads(text_area.toPlainText())
```
Please avoid the use of `_`, it is reserved for gettextization (see https://github.com/metabrainz/picard/blob/master/picard/i18n.py)

zas commented on plugins/album_level_tags/__init__.py:
```diff
@@ -0,0 +1,360 @@
+)
+from PyQt5 import QtGui
+from PyQt5.QtWidgets import QButtonGroup
+import time
```
You might want to re-order imports to fit usual Picard import order (see https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md and search for `isort`).

Sophist-UK commented:
I can't help but wonder whether it might be better for this plugin simply to calculate the lists and averages of various file metadata and place them in the album metadata as new variables.

Harakku commented:
Closing. It looks like these features were made available in the "Release tag aggregation functions" plugin: https://github.com/metabrainz/picard-plugins/pull/300
